### DISCUSSION TITLE
Browser Object Model

### DISCUSSION THREAD

**Discussion by mownica-webdev** [mownica-webdev/frontend_2024#25] [Tasks]
### **Browser Object Model Types:**

- Window Object Model: Represents the browser window and serves as the main tool for interaction within the Browser Object Model
- Location Object Model: Manages the current web address (URL) and allows changes within the Browser Object Model (BOM).
- Screen Object Model: Offers information about the user’s screen, like its size, within the Browser Object Model (BOM).
- History Object Model: Enables navigation control by keeping track of the user’s browsing history in the Browser Object Model (BOM).
- Navigator Object Model: Provides details about the user’s browser and system characteristics within the Browser Object Model (BOM).

### **1. Window Object Model:**
**Window Size:**

- window.innerHeight
- window.innerWidth

**Window Methods:**

- window.open() - This method is used to open a new tab or window with the specified URL and name.
- window.close() - This method is used for closing a certain window or tab of the browser that was previously opened by the
- window.open( ) method.
- window.moveTo() - This method is used in the window to move the window from the left and top coordinates.
- window.resizeTo() - This method is used to resize a window to the specified width and height

### **2. Location Object Model:**

- window.location.href: returns the href (URL) of the current page
- window.location.hostname: returns the domain name of the web host
- window.location.pathname: returns the path and filename of the current page
- window.location.protocol: returns the web protocol used (http: or https:)
- window.location.assign(): loads a new document

**3. Screen Object Model:**
**Window Screen Properties:**

screen.width
screen.height
screen.availWidth
screen.availHeight
screen.colorDepth
screen.pixelDepth

**4. History Object Model:**
**history object Properties:**

- length: It returns the length of the history URLs visited by the user in that session.
- history object Methods:
- 
- forward(): It loads the next page. Provides the same effect as clicking back in the browser.
- back(): It loads the previous page. Provides the same effect as clicking forward in the browser.
- go(): It loads the given page number in the browser. history.go(distance) function provides the same effect as pressing the back or forward button in your browser and specifying the page exactly that you want to load.

### **5. Navigator Object Model:**
**Navigator Object Properties:**

- navigator.userAgent: Returns the user agent string representing the browser's identification information.
- navigator.appName: Returns the official name of the browser.
- navigator.appVersion: Returns version information about the browser.
- navigator.platform: Returns the platform on which the browser is running (e.g., "Win32", "Linux", "MacIntel").
- navigator.language: Returns the language of the browser.
- navigator.cookieEnabled: Returns a boolean value indicating whether cookies are enabled in the browser.
- navigator.onLine: Returns a boolean value indicating whether the browser is currently online.
- navigator.plugins: Returns an array of Plugin objects representing the plugins installed in the browser.
- navigator.mimeTypes: Returns an array of MimeType objects representing the MIME types supported by the browser.
- navigator.hardwareConcurrency: Returns the number of logical processor cores available in the user's device.
- navigator.userAgentData: Returns information about the browser's user agent.

**Here's an example to access properties of the Navigator object:**

```
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Navigator Object Example</title>
  </head>
  <body>
    <script>
      // Getting user agent information
      const userAgent = navigator.userAgent;
      console.log("User Agent:", userAgent);

      // Getting the browser's official name
      const appName = navigator.appName;
      console.log("Browser Name:", appName);

      // Getting the browser's version information
      const appVersion = navigator.appVersion;
      console.log("Browser Version:", appVersion);

      // Getting the platform on which the browser is running
      const platform = navigator.platform;
      console.log("Platform:", platform);

      // Getting the language of the browser
      const language = navigator.language;
      console.log("Language:", language);

      // Checking if cookies are enabled in the browser
      const cookiesEnabled = navigator.cookieEnabled;
      console.log("Cookies Enabled:", cookiesEnabled);

      // Checking if the browser is currently online
      const isOnline = navigator.onLine;
      console.log("Online Status:", isOnline);

      // Getting the number of logical processor cores available in the user's device
      const hardwareConcurrency = navigator.hardwareConcurrency;
      console.log("Number of Cores:", hardwareConcurrency);

      // Accessing information about the browser's user agent
      const userAgentData = navigator.userAgentData;
      console.log("User Agent Data:", userAgentData);
    </script>
  </body>
</html>
```
### **JavaScript Timing Events**
**Timing Events:**
The window object allows execution of code at specified time intervals. These time intervals are called timing events.

- setTimeout() Method: Executes a function, after waiting a specified number of milliseconds.
- clearTimeout() method: stops the execution of the function specified in setTimeout().
- setInterval() method repeats a given function at every given time-interval.
- clearInterval() method stops the executions of the function specified in the setInterval() method.

**Comment by RAKESHSARIPINENI9490** [comment]
```
// Getting user agent information
    const userAgent = navigator.userAgent;
    console.log("User Agent:", userAgent);
 
    // Getting the browser's official name
    const appName = navigator.appName;
    console.log("Browser Name:", appName);
 
    // Getting the browser's version information
    const appVersion = navigator.appVersion;
    console.log("Browser Version:", appVersion);
 
    // Getting the platform on which the browser is running
    const platform = navigator.platform;
    console.log("Platform:", platform);
 
    // Getting the language of the browser
    const language = navigator.language;
    console.log("Language:", language);
 
    // Checking if cookies are enabled in the browser
    const cookiesEnabled = navigator.cookieEnabled;
    console.log("Cookies Enabled:", cookiesEnabled);
 
    // Checking if the browser is currently online
const isOnline = navigator.onLine;
    console.log("Online Status:", isOnline);
 
    // Getting the number of logical processor cores available in the user's device
    const hardwareConcurrency = navigator.hardwareConcurrency;
    console.log("Number of Cores:", hardwareConcurrency);
 
    // Accessing information about the browser's user agent
    const userAgentData = navigator.userAgentData;
    console.log("User Agent Data:", userAgentData);
 
    // Time Events
    console.log("Time Events:");
 
    // setTimeout() method
    setTimeout(() => {
      console.log("Timeout executed after 3 seconds!");
    }, 3000);
 
    // setInterval() method
    let intervalCount = 0;
    const intervalId = setInterval(() => {
      console.log(`Interval executed! Count: ${intervalCount}`);
      intervalCount++;
 
      // Clear interval 
      if (intervalCount === 10) {
        clearInterval(intervalId);
      }
    }, 1000);
 
```
   
<img width="724" alt="1" src="https://github.com/user-attachments/assets/c245c07f-60b8-4e8f-9daa-a0fe24bf9943">
